### DISCUSSION TITLE
ImportAction in combination with multi tenancy

### DISCUSSION THREAD

**Discussion by maartenpaauw** [filamentphp/filament#9982] [Help]
Today I've upgraded my side project to Filament V3.1 to toy around with the new ImportAction. The application uses the build in multi tenacy (tenant model named Organization) and a custom auth guard named 'organizer' (with auth model Organizer). With the application organizers within an organization can manage Spots. The idea is to make these spots importable using the new ImportAction.

First error I encountered was: `Cannot use “::class” on value of type null`. I'd had to add `$this->app->bind(Authenticatable::class, Organizer::class);` to a service provider, otherwise the `Import` model could not resolve the "user". It doesn’t feel like a good solution or am I wrong? Is this intended?

Secondly, The imported spots needs to be associated with the tenant (Organization). I've tried to receive to tenant using `Filament::tenant()` within the `resolveRecord` method, but this resolves to NULL. I think the reason is I'm using Horizon (Redis) as queue and `Filament::tenant()` only works on request lifecycle. How can I associate the current tenant to an imported record?

**Comment by maartenpaauw** [comment]
Today I came up with this solution. I had to use a hidden form element to store the tenant ID, because the `resolveRecord` method is called when the job is executed on the queue. If I'd call `Filament::getTenant()` within the `resolveRecord` method it resolves to `NULL`.

```php
public static function getOptionsFormComponents(): array
{
    return [
        Hidden::make('organization_id')
            ->default(Filament::getTenant()->getKey()),
    ];
}

public function resolveRecord(): ?Spot
{
    return new Spot(Arr::only($this->options, ['organization_id']));
}
```

**Comment by maartenpaauw** [comment]
Both solutions still feels like temporary fixes. In my opinion both things (resolving the user and tenant) should work out-of-the-box.

**Reply by huge10** [reply]
Struggling with the same issue regarding multi tenancy. Nice workaround but agreed this could be handled more elegantly.

**Reply by GerhardKotze** [reply]
Have you seen
https://github.com/filamentphp/filament/discussions/11224

**Comment by blenderdeluxe** [comment]
I know this thread is old, but I’ll share my experience from yesterday with my system in Filament in case my solution helps someone.

**Context**
Company Model
Worker Model
Even though I don’t use Filament’s multi-tenancy system, all queries are limited by a scope, so each user only sees the content of their company.

At first, I had trouble assigning the **company_id** to the worker. I retrieved it from **auth()->user()->company_id** and assigned it in the WorkerObserver. However, when using imports, a Job is executed, so I couldn’t access it.

I tried using lifehooks since the user is stored in the Job with Import $import->user(), but I couldn’t get it to work (maybe due to my lack of knowledge or the fact that it was 5 AM, and I had to deliver progress the next morning).

In the end, my solution was to add options in the ImportAction of my WorkerResource.

```
->headerActions([
                ImportAction::make()
                    ->importer(WorkerImporter::class)
                    ->options(['companyId' => auth()->user()->company_id])
                    ->maxRows(10000)
            ])
```

and then in my **WorkerImporter** I use the lifehook **beforeFill()**

```
protected function beforeFill(): void {
      if (!isset($this->data['company_id'])) {
          $this->data['company_id'] = $this->options['companyId'];
      }
}
```
And my WorkerObserver also I use

```
public function creating(Worker $worker): void
{
    if (is_null($worker->company_id)) {
        $worker->company_id = auth()->user()->company_id ?? null;
    }
}
```

I used this because I also run seeders locally to test and populate data.
I hope this will be useful for someone or even for me in the future hahaha. Of course, if there’s a better option, anyone is welcome to share!

**Comment by ali-erad** [comment]
I had a similar problem. I wanted to let the tenant user import products that are tenant aware so I had to find a way to "inject" the tenant_id into each record. Also, I wanted the relationships to be tenant aware to avoid data leak (allowing tenant 1 to use tenant 2's brand).

First, we "inject" the tenant ID into each to-be-imported-record:
```php
protected function beforeValidate(): void
{
    $this->record->tenant_id = Filament::getTenant()->id;
}
```

Doing this, tenant_id will be filled for each record. In the system, each product can belong to a brand. The brand is nullable.
```php
ImportColumn::make('brand')
    ->relationship(resolveUsing: function($state, ?SalesProduct $record, array $data) {
       //return null since it's nullable if the user didn't provide anything
        if(empty($state)) {
            return null;
        }

        /*
         * resolving the brand by name or id. The thing to note here is the use of 
         * $record to get the tenant_id since it's available before validtion.
         * and of course, correctly scope the query
         */
        $brand = Brand::query()
            ->where('tenant_id', $record->tenant_id)
            ->where(function ($query) use ($state, $data) {
                $query->where('name', $state)
                        ->orWhere('id', $state);
            })
            ->first();

        // proactively creating a new record if brand doesn't exists
        if(!$brand) {
            return Brand::create([
                'tenant_id' => $record->tenant_id,
                'name' => $data['brand'],
            ]);
        }

        return $brand;
    })
    ->ignoreBlankState()
    ->rules(['max:255'])
    ->example(__('Apple'))
    ->helperText(__('Can be resolved using name or system ID. if the brand doesn\'t exists, a new brand will be created for you')),
```

Hope this helps.
